offset_manager.go

```diff
@@ -251,28 +251,52 @@ func (om *offsetManager) Commit() {
 }
 
 func (om *offsetManager) flushToBroker() {
+	broker, err := om.coordinator()
+	if err != nil {
+		om.handleError(err)
+		return
+	}
+
+	// Care needs to be taken to unlock this. Don't want to defer the unlock as this would
+	// cause the lock to be held while waiting for the broker to reply.
+	broker.lock.Lock()
 	req := om.constructRequest()
 	if req == nil {
+		broker.lock.Unlock()
 		return
 	}
+	resp, rp, err := sendOffsetCommit(broker, req)
+	broker.lock.Unlock()
 
-	broker, err := om.coordinator()
 	if err != nil {
 		om.handleError(err)
+		om.releaseCoordinator(broker)
+		_ = broker.Close()
 		return
 	}
 
-	resp, err := broker.CommitOffset(req)
+	err = handleResponsePromise(req, resp, rp, nil)
 	if err != nil {
 		om.handleError(err)
 		om.releaseCoordinator(broker)
 		_ = broker.Close()
 		return
 	}
 
+	broker.handleThrottledResponse(resp)
 	om.handleResponse(broker, req, resp)
 }
 
+func sendOffsetCommit(coordinator *Broker, req *OffsetCommitRequest) (*OffsetCommitResponse, *responsePromise, error) {
+	resp := new(OffsetCommitResponse)
+	responseHeaderVersion := resp.headerVersion()
+	promise, err := coordinator.send(req, true, responseHeaderVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+	return resp, promise, err
```

Comment:
`err` here must be `nil`. It’s usually better to use the stronger statement. (That is, the value must be `nil` so making it `nil` allows readers to not have to consider what value `err` might have here.)

The alternative is to just replace the whole `if err != nil { return nil, nil, err } return resp, promise, err` with `return resp, promise, err`.

Reply:
Thanks for the careful review @puellanivis. I've added a further commit to based on your first suggestion.

I prefer keeping the if statement, as the alternative (`return resp, promise, err`) would return a non-nil `resp` value in the error case - which didn't seem quite right.

Reply:
👍 Yeah, I agree on keeping the `if` for the same reasons. Returning a non-nil `resp` in that case would be unexpected.